File: src/main/java/no/novari/flyt/archive/gateway/dispatch/journalpost/result/RecordDispatchResult.java
package no.novari.flyt.archive.gateway.dispatch.journalpost.result;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import no.novari.flyt.archive.gateway.dispatch.DispatchStatus;

@Getter
@ToString
@EqualsAndHashCode
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class RecordDispatchResult {

    public static RecordDispatchResult accepted(Long journalpostId) {
        return new RecordDispatchResult(DispatchStatus.ACCEPTED, journalpostId, null);
    }

    public static RecordDispatchResult declined(String errorMessage) {
        return new RecordDispatchResult(DispatchStatus.DECLINED, null, errorMessage);
    }

    public static RecordDispatchResult failed(String errorMessage) {
        return new RecordDispatchResult(DispatchStatus.FAILED, null, errorMessage);
    }

    public static RecordDispatchResult timedOut() {
        return new RecordDispatchResult(DispatchStatus.FAILED, null,
                "Record dispatch timed out. No response from destination."
        );
    }

    private final DispatchStatus status;
    private final Long journalpostId;
    private final String errorMessage;


}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/journalpost/result/RecordDispatchResult.kt
package no.novari.flyt.archive.gateway.dispatch.journalpost.result

import no.novari.flyt.archive.gateway.dispatch.DispatchStatus
import kotlin.ConsistentCopyVisibility

@ConsistentCopyVisibility
data class RecordDispatchResult private constructor(
    val status: DispatchStatus,
    val journalpostId: Long? = null,
    val errorMessage: String? = null,
) {
    companion object {
        @JvmStatic
        fun accepted(journalpostId: Long) = RecordDispatchResult(DispatchStatus.ACCEPTED, journalpostId = journalpostId)

        @JvmStatic
        fun declined(errorMessage: String) = RecordDispatchResult(DispatchStatus.DECLINED, errorMessage = errorMessage)

        @JvmStatic
        fun failed(errorMessage: String) = RecordDispatchResult(DispatchStatus.FAILED, errorMessage = errorMessage)

        @JvmStatic
        fun timedOut() =
            RecordDispatchResult(
                DispatchStatus.FAILED,
                errorMessage = "Record dispatch timed out. No response from destination.",
            )
    }
}
